src/plan_expand_hypertable.c

```diff
@@ -80,100 +80,95 @@ is_chunk_exclusion_func(Node *node)
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("only one chunks_in call is allowed per hypertable")));
+
+			Assert(func_expr->args->length == 2);
+			if (!IsA(linitial(func_expr->args), Var))
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("first parameter for chunks_in function needs to be a record")));
+
+			ctx->chunk_exclusion_func = func_expr;
+			quals = (Node *) list_delete_cell((List *) quals, lc, prev);
+			lc = prev;
+		}
+		else
 		{
-			Node *qual = (Node *) lfirst(lc);
-			RestrictInfo *restrictinfo;
-			Relids relids;
-			bool belongs_to_cur_rel = false;
-
-			next = lnext(lc);
-
-			relids = pull_varnos(qual);
-
-			belongs_to_cur_rel =
-				bms_num_members(relids) == 1 && bms_is_member(ctx->rel->relid, relids);
-
-			if (!belongs_to_cur_rel)
-			{
-				lc = next;
-				continue;
-			}
-
-			if (is_chunk_exclusion_func(qual))
-			{
-				FuncExpr *func_expr = (FuncExpr *) qual;
-
-				/* validation */
-				if (ctx->chunk_exclusion_func != NULL)
-					ereport(ERROR,
-							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-							 errmsg("only one chunks_in call is allowed per hypertable")));
-
-				Assert(func_expr->args->length == 2);
-				if (!IsA(linitial(func_expr->args), Var))
-					ereport(ERROR,
-							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-							 errmsg(
-								 "first parameter for chunks_in function needs to be a record")));
-
-				ctx->chunk_exclusion_func = func_expr;
-			}
-
-			if (ctx->chunk_exclusion_func == NULL)
-			{
 #if PG96
-				restrictinfo =
-					make_restrictinfo((Expr *) qual, true, false, false, relids, NULL, NULL);
+			restrictinfo = make_restrictinfo((Expr *) qual, true, false, false, relids, NULL, NULL);
 #else
-				restrictinfo = make_restrictinfo((Expr *) qual,
-												 true,
-												 false,
-												 false,
-												 ctx->root->qual_security_level,
-												 relids,
-												 NULL,
-												 NULL);
+			restrictinfo = make_restrictinfo((Expr *) qual,
+											 true,
+											 false,
+											 false,
+											 ctx->root->qual_security_level,
+											 relids,
+											 NULL,
+											 NULL);
 #endif
-				ctx->restrictions = lappend(ctx->restrictions, restrictinfo);
-			}
-			else if (!func_removed && ctx->chunk_exclusion_func != NULL)
-			{
-				f->quals = (Node *) list_delete_cell((List *) (f->quals), lc, prev);
-				func_removed = true;
-			}
-			prev = lc;
-			lc = next;
+			ctx->restrictions = lappend(ctx->restrictions, restrictinfo);
 		}
-		if (ctx->chunk_exclusion_func != NULL)
-			return (Node *) f;
+	}
+	return quals;
+}
+
+static bool
+collect_quals_walker(Node *node, CollectQualCtx *ctx)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, FromExpr))
+	{
+		FromExpr *f = castNode(FromExpr, node);
+		f->quals = process_quals(f->quals, ctx);
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *j = castNode(JoinExpr, node);
+		j->quals = process_quals(j->quals, ctx);
 	}
 
```

Comment:
(optional) add `if (ctx->chunk_exclusion_func != NULL) return node;` to short-cut for efficiency?

Reply:
we still need to process all quals to detect duplicate chunks_in calls